### ISSUE TITLE
xcode 15 and ocamlmklib 4.14.1 warning on c stubs generation

### ISSUE THREAD

**Issue by dbuenzli** [ocaml/ocaml#12641]
For some reason I got tricked into updating xcode and now `ocamlmklib` spits the following annoying warning when generating c stubs:

```
> '/Users/dbuenzli/.opam/4.14.1/bin/ocamlmklib' '-v' '-v' \ 
  '-o' '…/_b0/b/user/webs-unix/webs_unix_stubs' '-g' \
  '…/_b0/b/user/webs-unix/webs_stubs_time.o' '…/_b0/b/user/webs-unix/webs_stubs_sendfile.o'
+ cc -shared                    -flat_namespace -undefined suppress -Wl,-no_compact_unwind                     -g -o …/_b0/b/user/webs-unix/dllwebs_unix_stubs.so …/_b0/b/user/webs-unix/webs_stubs_time.o …/_b0/b/user/webs-unix/webs_stubs_sendfile.o     
ld: warning: -undefined suppress is deprecated
ld: warning: -undefined suppress is deprecated
+ ar rcs …/_b0/b/user/webs-unix/libwebs_unix_stubs.a  …/_b0/b/user/webs-unix/webs_stubs_time.o …/_b0/b/user/webs-unix/webs_stubs_sendfile.o
```

Maybe https://github.com/ocaml/ocaml/issues/11915 solves this ?

However FWIW I tried with `LDFLAGS=-w opam reinstall ocaml-base-compiler`, but it seems I still get 1 warning (rather than two above), the resulting invocation is:
```
+ cc -shared                    -flat_namespace -undefined suppress -Wl,-no_compact_unwind                    -w -g -o …/_b0/b/user/webs-unix/dllwebs_unix_stubs.so …/_b0/b/user/webs-unix/webs_stubs_time.o …/_b0/b/user/webs-unix/webs_stubs_sendfile.o     
ld: warning: -undefined suppress is deprecated
```

**Comment by johnwhitington** [comment]
I get this in 4.14 like you, but not in 5.1.0.

**Comment by dbuenzli** [comment]
It seems that 5.1.0 is using `-undefined dynamic_lookup` which doesn't warn on macos arm64 (but might elsewhere as already noted by @xavierleroy [here](https://github.com/ocaml/ocaml/pull/11571#issuecomment-1259170740)).

Also xref https://github.com/coq/coq/issues/18134 /cc @pmetzger

**Comment by pmetzger** [comment]
@dbuenzli Thanks for the ping.

**Comment by dbuenzli** [comment]
After two days of getting this warning in my face I'm getting fairly annoyed. 

I cargo culted the 5.1.0 changes [here](https://github.com/dbuenzli/ocaml/commit/1751abfa717b9ec21c294dc95fda4d132e2eb22d) should we perhaps patch the tarball in the opam repository with it ?

**Comment by pmetzger** [comment]
@dbuenzli I may add your patch to MacPorts, at least for XCode 15. It would seem to do the correct thing.

**Comment by pmetzger** [comment]
@dbuenzli Unfortunately, when I add your patch, Coq will no longer build. I end up with errors like:

```
Error: I/O error: dllzarith.so: No such file or directory
```

You might want to try building Coq 8.18.0 and see if you experience a similar error.

(I worry that this might be spurious and the problem might be that I need to rebuild zarith etc. from scratch. I'll be trying that soon.)

(Amended to add: I rebuilt zarith and reinstalled it with the modified compiler, and it didn't fix anything.)

**Comment by dbuenzli** [comment]
Note I just tested that on a build directly from the tarball, it seems to work building the compiler but I didn't test it on an opam install. 

I'm not even sure you can change these parameters without changing the way the code is generated – note the "cargo cult" bit in my message. 

I just rolled a ball hoping someone who knows can confirm or disprove the correctness of the patch.

**Comment by pmetzger** [comment]
I built the compiler without any trouble with your patch, and a few simple packages, but Coq seems to break it.

**Comment by dbuenzli** [comment]
Could someone help with this ? I'm happy to do the work if I'm given directions. 

It's genuinely annoying for users and I'm not ready to use OCaml > 5 for now.

**Comment by Octachron** [comment]
Once OCaml 5.1.1 is out (soonish), I will try to focus on OCaml 4.14.2 and backporting fixes.

In the meanwhile, I will try to have a branch with the changes on the macOS linker ready this week if you want to test if those work.

**Comment by dbuenzli** [comment]
Thanks @Octachron! All my testing belongs to yours.

**Comment by Octachron** [comment]
Test in progress at https://github.com/Octachron/ocaml/tree/4.14_macOS_linking (which is simplified version of your changes).

**Comment by dbuenzli** [comment]
Is there an easy way to pin the compiler to this ? I tried with: 

```
opam pin add ocaml-base-compiler.4.14.1 git@github.com:Octachron/ocaml.git#4.14_macOS_linking
```
On my work switch but that eventually led to: 

```
[ERROR] The compilation of ocaml.4.14.1 failed at "ocaml /Users/dbuenzli/.opam/4.14.1/share/ocaml-config/gen_ocaml_config.ml 4.14.1 ocaml".

#=== ERROR while compiling ocaml.4.14.1 =======================================#
# context     2.1.5 | macos/arm64 | ocaml-base-compiler.4.14.1 | https://opam.ocaml.org#b61304c6
# path        ~/.opam/4.14.1/.opam-switch/build/ocaml.4.14.1
# command     ~/.opam/opam-init/hooks/sandbox.sh build ocaml /Users/dbuenzli/.opam/4.14.1/share/ocaml-config/gen_ocaml_config.ml 4.14.1 ocaml
# exit-code   1
# env-file    ~/.opam/log/ocaml-22162-6d9079.env
# output-file ~/.opam/log/ocaml-22162-6d9079.out
### output ###
# OCaml version mismatch: 4.14.2, expected 4.14.1
```

Which is a pity, is there a way to bypass this test ? 

If I try to pin to `ocaml-base-compiler.4.14.2` that first gets you in an elaborate interactive mode which eventually fails because (I guess) of switch invariant:

```
[ERROR] Package conflict!
  * Missing dependency:
    - ocaml-base-compiler = 4.14.1 | ocaml-system = 4.14.1
    not available because the package is pinned to version 4.14.2
    unmet availability conditions: 'sys-ocaml-version = "4.14.1"'
```

At least I got to exercise `opam`'s `install --restore` path, which I have to say works like a charm.

**Comment by lthls** [comment]
I usually pin `ocaml-variants` instead of `ocaml-base-compiler` when testing compiler branches. Something like that should work:
```
opam pin add ocaml-variants.4.14.2 git@github.com:Octachron/ocaml.git#4.14_macOS_linking
```

**Comment by Octachron** [comment]
Don't you need to create a switch first? Something like
```bash
opam switch create 4.14.2+test --empty
opam pin add ocaml-variants.4.14.2 git@github.com:Octachron/ocaml.git#4.14_macOS_linking
```
works for me.

**Comment by dra27** [comment]
You forgot your #9298, @dbuenzli! 🙂 The compiler's `ocaml-variants.opam` file always contains an accurate version number (it was added to the release checklist in #8604), so one can save a few keystrokes by omitting the version number and just do `opam pin add ocaml-variants <url>`

**Comment by dbuenzli** [comment]
> Don't you need to create a switch first?

Sure I can do that. But I simply wanted to test it on my current work base. That's the beauty of pins !

**Comment by dbuenzli** [comment]
So to sum up that's the invocation I was after:

```
opam pin add ocaml-variants --update-invariant git@github.com:Octachron/ocaml.git#4.14_macOS_linking 
```

No warnings, and things seem to work.

**Comment by Octachron** [comment]
I have pushed the change on the 4.14 branch, I will let this issue opens until the 4.14.2 release candidate to remind myself to look for potential macOS issues.

**Comment by pmetzger** [comment]
So I can add a patch for MacPorts, can someone point me to the correct diff I should apply on the unpacked OCaml sources?

**Comment by Octachron** [comment]
This specific fix is now available in 4.14.2 . I will try to have a more timely 4.14.3 release is other issues arise.

**Comment by pmetzger** [comment]
@Octachron Thank you! This was helpful!
